File: python/metiquo/providers/stake_dom.py
"""Extraction du DOM public Stake : aucune lecture du store JS ou du trafic réseau."""

EXTRACT_DOM = r"""() => {
  const root = document.querySelector('#main-content');
  if (!root) throw new Error('STAKE_DOM_MISSING');
  const clean = value => (value || '').replace(/\s+/g, ' ').trim();
  const visible = node => node.getClientRects().length > 0;
  const time = root.querySelector('[data-table="time"]');
  const parentPath = location.pathname.replace(/\/$/, '').split('/').slice(0, -1).join('/');
  const competition = Array.from(document.querySelectorAll('a[href]'))
    .find(a => a.getAttribute('href') === parentPath);
  const participants = Array.from(root.querySelectorAll('[data-testid="competitor-item"]'))
    .map(n => clean(n.closest('[role="group"]')?.getAttribute('aria-label'))).filter(Boolean);
  return {
    url: location.origin + location.pathname,
    competition: clean(competition?.textContent),
    participants: [...new Set(participants)],
    displayedStart: clean(time?.parentElement?.textContent),
    scores: Array.from(root.querySelectorAll('[data-testid="score-ticker-item"]'))
      .map(n => clean(n.textContent)),
    tabs: Array.from(root.querySelectorAll('button[data-testid^="tab-"]'))
      .map(n => n.getAttribute('data-testid')),
    links: Array.from(root.querySelectorAll('a[href]'))
      .map(n => n.getAttribute('href')),
    markets: Array.from(root.querySelectorAll('.secondary-accordion.level-2'))
      .filter(visible).map(n => ({
        label: clean(n.querySelector(':scope > .header [data-ds-text]')?.textContent),
        expanded: n.classList.contains('is-open'),
        rawText: clean(n.textContent),
        outcomes: Array.from(n.querySelectorAll('[data-testid="fixture-outcome"]'))
          .filter(visible).map(b => ({
            label: clean(b.getAttribute('aria-label')),
            displayedLabel: clean(
              b.querySelector('[data-testid="outcome-button-name"]')?.textContent),
            oddsText: clean(b.querySelector('[data-testid="fixture-odds"]')?.textContent),
            disabled: b.disabled || b.getAttribute('aria-disabled') === 'true',
            text: clean(b.textContent),
          })),
      })),
  };
}"""

File: apps/worker/src/metiquo_worker/stake_dom.js
// Read-only snapshot for the manual Stake feasibility audit. Never clicks a selection.
() => {
  const contentRoot = document.querySelector('#main-content');
  const root = contentRoot ?? document.createElement('div');
  const text = (node) => node?.textContent?.trim().replace(/\s+/g, ' ') ?? '';
  const url = (value) => {
    try {
      const parsed = new URL(value, location.href);
      return parsed.origin + parsed.pathname;
    } catch {
      return null;
    }
  };
  const rect = (node) => {
    const r = node.getBoundingClientRect();
    return { x: r.x, y: r.y, width: r.width, height: r.height };
  };
  const links = (parent) =>
    [...parent.querySelectorAll('a[href]')]
      .filter((a) => /\/sports\/(esports|league-of-legends)(\/|$)/.test(a.href))
      .map((a) => ({ text: text(a), url: url(a.href) }));
  const number = (raw) => (/^[+-]?\d+(?:[.,]\d+)?$/.test(raw) ? raw.replace(',', '.') : null);
  const selections = (parent, title) =>
    [...parent.querySelectorAll('[data-testid="fixture-outcome"]')].map((button, index) => {
      const name = text(button.querySelector('[data-testid="outcome-button-name"]'));
      const odds = text(button.querySelector('[data-testid="fixture-odds"]'));
      const table = button.closest('.table-columns');
      const column = button.closest('.column');
      const columns = table
        ? [...table.children].filter((n) => n.classList.contains('column'))
        : [];
      const headings = columns.filter((n) => n.classList.contains('heading'));
      const cells = columns.filter((n) => !n.classList.contains('heading'));
      const position = cells.indexOf(column);
      const heading =
        headings.length && position >= 0 ? headings[position % headings.length] : null;
      const hasNumericLine =
        /handicap|nombre|total|durée|duration|premier.*atteindre|first.*to/i.test(title);
      return {
        index,
        nameRaw: name,
        accessibleName: button.getAttribute('aria-label'),
        columnHeaderRaw: heading ? text(heading) : null,
        lineRaw: hasNumericLine && number(name) !== null ? name : null,
        line: hasNumericLine ? number(name) : null,
        oddsRaw: odds || null,
        decimalOdds: number(odds),
        disabled: button.disabled,
        text: text(button),
        rect: rect(button),
      };
    });
  const markets = [...root.querySelectorAll('.groups .secondary-accordion')].map((group) => {
    const header = group.querySelector(':scope > .header');
    const title = text(header?.querySelector('span[data-ds-text]')) || text(header);
    return {
      title,
      headerRaw: text(header),
      expanded: group.classList.contains('is-open'),
      text: text(group),
      columnHeadersRaw: [...group.querySelectorAll('.table-columns .column.heading')].map(text),
      selections: selections(group, title),
      controls: [...group.querySelectorAll('button:not([data-testid="fixture-outcome"])')].map(
        (b) => ({
          text: text(b),
          testId: b.getAttribute('data-testid'),
          disabled: b.disabled,
        }),
      ),
    };
  });
  const fixtures = [...root.querySelectorAll('[data-testid="fixture-preview"]')].map((fixture) => {
    const observedLinks = links(fixture);
    const eventLink = observedLinks.find((link) =>
      /\/sports\/league-of-legends\/[^/]+\/[^/]+\/\d+-[^/]+$/.test(link.url ?? ''),
    );
    const eventUrl = eventLink?.url;
    const parentUrl = eventUrl?.slice(0, eventUrl.lastIndexOf('/'));
    const categoryUrl = parentUrl?.slice(0, parentUrl.lastIndexOf('/'));
    return {
      event: eventLink
        ? { ...eventLink, idFromUrl: eventUrl.match(/\/(\d+)-[^/]+$/)?.[1] ?? null }
        : null,
      competition: observedLinks.find((link) => link.url === parentUrl) ?? null,
      category: observedLinks.find((link) => link.url === categoryUrl) ?? null,
      text: text(fixture),
      links: observedLinks,
      selections: selections(fixture, ''),
    };
  });
  return {
    contentRootFound: contentRoot !== null,
    url: url(location.href),
    title: document.title,
    viewport: { width: innerWidth, height: innerHeight },
    mainText: text(root),
    links: links(document),
    fixtures,
    tabs: [...root.querySelectorAll('.groups button[data-testid^="tab-"]')].map((b) => ({
      testId: b.getAttribute('data-testid'),
      text: text(b),
      classes: b.className,
      disabled: b.disabled,
    })),
    markets,
    listingControls: [...root.querySelectorAll('button')]
      .filter((b) => !b.closest('.groups') && !b.matches('[data-testid="fixture-outcome"]'))
      .map((b) => ({
        text: text(b),
        testId: b.getAttribute('data-testid'),
        label: b.getAttribute('aria-label'),
        disabled: b.disabled,
      })),
  };
};
